fem/bilinearform.hpp

```diff
@@ -586,21 +636,43 @@ class MixedBilinearForm : public Matrix
    /// Adds a boundary integrator. Assumes ownership of @a bfi.
    void AddBoundaryIntegrator(BilinearFormIntegrator *bfi);
 
+   /// Adds a boundary integrator. Assumes ownership of @a bfi.
+   void AddBoundaryIntegrator (BilinearFormIntegrator * bfi,
+                               Array<int> &bdr_marker);
+
    /** @brief Add a trace face integrator. Assumes ownership of @a bfi.
 
        This type of integrator assembles terms over all faces of the mesh using
        the face FE from the trial space and the two adjacent volume FEs from the
        test space. */
    void AddTraceFaceIntegrator(BilinearFormIntegrator *bfi);
 
+   /// Adds a boundary trace face integrator. Assumes ownership of @a bfi.
+   void AddBdrTraceFaceIntegrator (BilinearFormIntegrator * bfi);
```

Comment:
Hi @najlkin, I'm wondering -- are you actually using this type of integrators? The `TraceFaceIntegrator`s arise in DPG discretizations, so I'm curious if you are using the boundary version also for DPG.

Reply:
Hi @v-dobrev , I'm using these with mixed elements. Just consider formulation of weak gradient with L2 + Hdiv spaces. When considering general case with non-zero values at the boundaries, there arise boundary integrals, which should be correctly implemented this way (or you may incorrectly put "frozen" values of L2 potentials to RHS by using `BoundaryLFIntegrator` with `GridFunctionBdrTraceCoefficient` from #716 ). There is a well-known relation between mixed and DG methods, but I personally prefer dual formulation with fluxes and hybridization eventually, because of exact potentials conservation. Btw, it reminds me that I could make a pull request for `AddBdrConstraintIntegrator` in `Hybridization` to deal with BCs for fluxes, if you are not against that...

Reply:
I don't quite understand what type of equation and BCs you are using this for -- is it the Laplace equation with Dirichlet BCs?

Anyway, it is not that important for me to know exactly how you are using this -- what matters is that you do need it.

Regarding extending the `Hybridization` class -- I'm not sure I understand why you need to add boundary constraints. If you want to impose essential BCs on the original system (before hybridization) you can already do that.

Reply:
Sry, I was not very clear on that. Let's say you have the Laplace equation, but you do NOT have Dirichlet for potentials ("temperature"), putting aside Ill-posedness (consider it is just on a part of the boundary or so). Applying Green's theorem, the weak gradient has the form -(T, div w) + <T, w.n>, where T is the "temperature" and w the flux test function. The boundary integral breaks the symmetry of the problem, so you may want to get around placing it directly into the system matrix, but what if you really want to put it there?
Regarding `Hybridization`, you are right that you may impose essential BCs by marking essential vdofs at the boundaries. What I meant was imposing the essential BCs weakly, i.e. in DG-like fashion, where you would use `AddBdrFaceIntegrator`. It is not necessary to add this feature, but it would be nice :wink: (and I have it in my fork already).